### PR TITLE
Fix compile with clang and sanitizers

### PR DESCRIPTION
**Pull request by proller**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -147,7 +147,7 @@ set (CMAKE_C_FLAGS                       "${CMAKE_C_FLAGS} ${COMPILER_FLAGS} -fn
 set (CMAKE_C_FLAGS_RELWITHDEBINFO        "${CMAKE_C_FLAGS_RELWITHDEBINFO} -O3 ${CMAKE_C_FLAGS_ADD}")
 set (CMAKE_C_FLAGS_DEBUG                 "${CMAKE_C_FLAGS_DEBUG} -O0 -g3 -ggdb3 -fno-inline ${CMAKE_C_FLAGS_ADD}")
 
-if (MAKE_STATIC_LIBRARIES AND NOT APPLE AND NOT (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND OS_FREEBSD))
+if (MAKE_STATIC_LIBRARIES AND NOT APPLE AND NOT (COMPILER_CLANG AND OS_FREEBSD) AND NOT (COMPILER_CLANG AND SANITIZE))
```
Comment (alexey-milovidov) [inline CMakeLists.txt]: It works fine on any machine I have tested. Why we need this condition?

Comment (proller) [reply CMakeLists.txt]: ubuntu bionic. clang-6 from package: 

```rm CMakeCache.txt  && cmake .. -DSANITIZE=address -DCMAKE_CXX_COMPILER=/usr/bin/clang\+\+-6.0 -DCMAKE_C_COMPILER=/usr/bin/clang-6.0 -GNinja && ninja```
```-- Performing Test HAVE_THREE_PARAM_SCHED_SETAFFINITY
-- Performing Test HAVE_THREE_PARAM_SCHED_SETAFFINITY - Failed
-- Performing Test HAVE_TWO_PARAM_SCHED_SETAFFINITY
-- Performing Test HAVE_TWO_PARAM_SCHED_SETAFFINITY - Failed```

```
[17/316] Building CXX object contrib/poco/Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.o
FAILED: contrib/poco/Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.o 
/usr/bin/ccache /usr/bin/clang++-6.0  -DHAVE_PTHREAD_SETAFFINITY_NP -DPCRE_STATIC -DPOCO_EXTERNAL_OPENSSL -DPOCO_HAVE_FD_EPOLL -DPOCO_STATIC -DPOCO_UNBUNDLED -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_REENTRANT -D_THREAD_SAFE -D_XOPEN_SOURCE=500 -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/Foundation/src -I../contrib/zlib-ng -Icontrib/zlib-ng -g -fno-omit-frame-pointer -DSANITIZER -fsanitize=address -fsanitize-address-use-after-scope -fdiagnostics-color=always  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall -Wno-unused-command-line-argument  -Wnon-virtual-dtor -stdlib=libc++ -Wno-old-style-cast -Wno-unused-function -Wno-unused-variable -Wno-unused-result -Wno-deprecated-declarations -Wno-non-virtual-dtor -Wno-format -std=c++1z -Wno-sign-compare  -fno-strict-aliasing -Wno-unknown-pragmas -Wno-non-virtual-dtor -O2 -g -DNDEBUG -O3    -std=c++1z -MD -MT contrib/poco/Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.o -MF contrib/poco/Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.o.d -o contrib/poco/Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.o -c ../contrib/poco/Foundation/src/Thread.cpp
distcc[673499] ERROR: compile ../contrib/poco/Foundation/src/Thread.cpp on localhost failed
In file included from ../contrib/poco/Foundation/src/Thread.cpp:21:
In file included from ../contrib/poco/Foundation/src/Thread_STD.cpp:25:
../contrib/poco/Foundation/src/Thread_STD_POSIX.cpp:170:6: error: no matching function for call to 'pthread_setaffinity_np'
        if (pthread_setaffinity_np(_pData->thread->native_handle(), &cpuset) != 0)
            ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/pthread.h:475:12: note: candidate function not viable: requires 3 arguments, but 2 were provided
extern int pthread_setaffinity_np (pthread_t __th, size_t __cpusetsize,
           ^
In file included from ../contrib/poco/Foundation/src/Thread.cpp:21:
In file included from ../contrib/poco/Foundation/src/Thread_STD.cpp:25:
../contrib/poco/Foundation/src/Thread_STD_POSIX.cpp:206:6: error: no matching function for call to 'pthread_getaffinity_np'
        if (pthread_getaffinity_np(_pData->thread->native_handle(), &cpuset) != 0)
            ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/pthread.h:480:12: note: candidate function not viable: requires 3 arguments, but 2 were provided
extern int pthread_getaffinity_np (pthread_t __th, size_t __cpusetsize,
           ^
2 errors generated.
```

CMakeFiles/CMakeError.log:
```
Performing C++ SOURCE FILE Test HAVE_THREE_PARAM_SCHED_SETAFFINITY failed with the following output:                                                                                                               
Change Dir: /opt/home/proller/ClickHouse5/build_test1/CMakeFiles/CMakeTmp                                                                                                                                          
                                                                                                                                                                                                                   
Run Build Command:"/usr/bin/ninja" "cmTC_77fa8"                                                                                                                                                                    
[1/2] Building CXX object CMakeFiles/cmTC_77fa8.dir/src.cxx.o                                                                                                                                                      
[2/2] Linking CXX executable cmTC_77fa8                                                                                                                                                                            
FAILED: cmTC_77fa8                                                                                                                                                                                                 
: && /usr/bin/clang++-6.0  -g -fno-omit-frame-pointer -DSANITIZER -fsanitize=address -fsanitize-address-use-after-scope -fdiagnostics-color=always  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wal
l -Wno-unused-command-line-argument  -Wnon-virtual-dtor -stdlib=libc++ -Wno-old-style-cast -Wno-unused-function -Wno-unused-variable -Wno-unused-result -Wno-deprecated-declarations -Wno-non-virtual-dtor -Wno-for
mat -std=c++1z -Wno-sign-compare  -fno-strict-aliasing -Wno-unknown-pragmas -Wno-non-virtual-dtor -DHAVE_THREE_PARAM_SCHED_SETAFFINITY  -fsanitize=address -fsanitize-address-use-after-scope -fuse-ld=lld -static-
libgcc -static-libstdc++   -rdynamic CMakeFiles/cmTC_77fa8.dir/src.cxx.o  -o cmTC_77fa8   && :                                                                                                                     
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__si_class_type_info                                                                                                               
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
                                                                                                                                                                                                                   
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__class_type_info                                                                                                                  
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
                                                                                                                                                                                                                   
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: __dynamic_cast                                                                                                                                              
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
                                                                                                                                                                                                                   
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__vmi_class_type_info                                                                                                              
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
                                                                                                                                                                                                                   
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: __dynamic_cast                                                                                                                                              
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
                                                                                                                                                                                                                   
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: __dynamic_cast                                                                                                                                              
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
 

Performing C++ SOURCE FILE Test HAVE_READLINE_HISTORY failed with the following output:                                                                                                                            
Change Dir: /opt/home/proller/ClickHouse5/build_test1/CMakeFiles/CMakeTmp                                                                                                                                          
                                                                                                                                                                                                                   
Run Build Command:"/usr/bin/ninja" "cmTC_58e0c"                                                                                                                                                                    
[1/2] Building CXX object CMakeFiles/cmTC_58e0c.dir/src.cxx.o                                                                                                                                                      
[2/2] Linking CXX executable cmTC_58e0c                                                                                                                                                                            
FAILED: cmTC_58e0c                                                                                                                                                                                                 
: && /usr/bin/clang++-6.0  -g -fno-omit-frame-pointer -DSANITIZER -fsanitize=address -fsanitize-address-use-after-scope -fdiagnostics-color=always  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wal
l -Wno-unused-command-line-argument  -Wnon-virtual-dtor -stdlib=libc++ -DHAVE_READLINE_HISTORY  -fsanitize=address -fsanitize-address-use-after-scope -fuse-ld=lld -static-libgcc -static-libstdc++   -rdynamic CMa
keFiles/cmTC_58e0c.dir/src.cxx.o  -o cmTC_58e0c  /usr/lib/x86_64-linux-gnu/libreadline.a /usr/lib/x86_64-linux-gnu/libtermcap.a && :                                                                               
/usr/lib/llvm-6.0/bin/ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__si_class_type_info                                                                                                               
>>> referenced by ubsan_type_hash_itanium.cc.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/lib/llvm-6.0/lib/clang/6.0.0/lib/linux/lib
clang_rt.asan_cxx-x86_64.a                                                                                                                                                                                         
```

Comment (alexey-milovidov) [reply CMakeLists.txt]: It should work fine with `libc++`.

You can modify your test to only disable static linking when libstdc++ is used (that's not typical when using clang).

**Comment by alexey-milovidov** [comment]
There are no differences between this PR and master.

**Pull request closed without merging**